### ISSUE TITLE
Implement Security in Resthooks

### ISSUE THREAD

**Issue by callmekatootie** [topcoder-platform/vorbote#7]
[This](http://resthooks.org/docs/security/)

**Comment by cwdcwd** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Implement Security in Resthooks](https://trello.com/c/QHOQhNS3/10-implement-security-in-resthooks)

**Comment by cwdcwd** [comment]
1. immediate, methinks
2. maybe we skip a complicated content filtering scheme and do skinny payloads for non-admin/copilot users; though we still might want to filter certain sensitive data regardless of role

**Comment by callmekatootie** [comment]
For 2, I think that we may have to discuss a bit more. If we are implementing skinny payloads, we need to store incoming messages while the hook will contain the url to access these messages. Perhaps we can take it up in a new ticket itself

For 1, let me know if my understanding of how to implement security is right - when user creates a webhook, vorbote will give them a unique key. Next, when vorbote notifies the hook, it passes this unique key in the header. The hook needs to respond back with the unique key. Is that right?

**Comment by cwdcwd** [comment]
1. when the user creates the hook, I believe what needs to happen is that Vorbote sends a confirmation POST back to the hook endpoint for confirmation with the `Hook-Secret` header. The endpoint should return that `Hook-Secret` header. What that does is prevents malicious users from creating webhooks that DOS endpoints. The endpoint needs to respond back saying "yes" before we start sending data to it. So an "activation" flag will be needed on the model for the registered hook. 

![image](https://user-images.githubusercontent.com/1504607/43805098-83dd815e-9a6c-11e8-8c67-88cc792b0d8e.png)

**Comment by cwdcwd** [comment]
2. let's open another ticket for discussion on skinny payloads

**Comment by callmekatootie** [comment]
1. Makes sense. I got confused here because we are creating the webhook through the UI and not through an endpoint but I now realize that we can send the secret _after_ the webhook is created by POSTing to the webhook
2. https://github.com/topcoder-platform/vorbote/issues/16 created for skinny payloads

**Comment by callmekatootie** [comment]
Taken up in contest http://www.topcoder.com/challenge-details/30101652/?type=develop

**Comment by callmekatootie** [comment]
This has now been resolved
